db/transactions.go

```diff
@@ -21,29 +21,36 @@ func NewTxManager(db *sqlx.DB) *TxManager {
 	return &TxManager{DB: db}
 }
 
-func (s *TxManager) WithTransaction(ctx context.Context, f func(context.Context, Querier) error) (err error) {
-	ctx, span := o11y.StartSpan(ctx, "tx-manager: with-transaction")
-	defer o11y.End(span, &err)
+func retryable(err error) bool {
+	return err != nil && errors.Is(err, ErrBadConn)
+}
 
+func (s *TxManager) WithTransaction(ctx context.Context, f func(context.Context, Querier) error) (err error) {
 	// Retry this transaction 3 times
 	for i := 0; i < 3; i++ {
-		err = s.WithOneTransaction(ctx, f)
-		if !errors.Is(err, ErrBadConn) {
+		err = s.WithOneTransaction(ctx, i, f)
+		if !retryable(err) {
 			break
 		}
-		o11y.AddField(ctx, "bad_con", i)
-		o11y.AddField(ctx, "warning", err)
 	}
 
 	// Note that the above defer can reassign err
 	return err
 }
 
-func (s *TxManager) WithOneTransaction(ctx context.Context, f func(context.Context, Querier) error) (err error) {
+func (s *TxManager) WithOneTransaction(ctx context.Context,
+	attempt int, f func(context.Context, Querier) error) (err error) {
 	ctx, span := o11y.StartSpan(ctx, "tx-manager: with-one-transaction")
-	defer o11y.End(span, &err)
-
+	defer func() {
+		if retryable(err) {
+			o11y.AddField(ctx, "bad_con", attempt)
+			o11y.AddField(ctx, "warning", err)
+		}
+		o11y.End(span, &err)
```

Comment:
I dont think this End now plays well with the warning above

Reply:
yeah don't need the warning

Reply:
hmm the intention was that this error (bad connection) can be considered a warning (we do not need to take any actions to reduce its occurrence) when it is retried.

If after 3 failed attempts then it should be logged as an error - e.g. if we keep seeing that error we should act.